File: javascript/03-advanced/01-prototypes/exercise.js
// exercise.js - Practice with Prototypes

// TODO 1: Create a 'parent' object with a property 'role' set to "user".

// TODO 2: Create a 'child' object and set its prototype to 'parent'.

// TODO 3: Log 'child.role'. Then change 'child.role' to "admin" and log it again.
// Verify if 'parent.role' changed or stayed the same.

// TODO 4: Use 'Object.create()' to create a 'square' object that 
// inherits from a 'shape' object (which has a 'type' property).

// TODO 5: Create a constructor function 'Person(name)'. 
// Add a 'greet()' method to 'Person.prototype'.

// TODO 6: Create two instances of 'Person'. 
// Use '=== ' to check if their 'greet' methods are the same exact function.
/*
const p1 = new Person("A");
const p2 = new Person("B");
console.log(p1.greet === p2.greet); // Should be true if using prototype!
*/

// CHALLENGE: Create a custom method 'last()' on 'Array.prototype' 
// that returns the last element of any array.
// [1, 2, 3].last(); // Expected: 3

// --- Verification ---
console.log("--- Results ---");
try {
    // console.log(`Method shared: ${p1.greet === p2.greet}`);
} catch (e) {
    console.log("Error: fill in all TODOs to see results!");
}

File: python/03-advanced/02-metaclasses/exercise.py
# exercise.py - Practice Metaclasses

# TODO 1: Create a class 'MyDynamicClass' dynamically using the 'type()' function.
# It should have one method 'hello()' that returns "Greetings!".
# MyDynamicClass = ...

# TODO 2: Write a metaclass 'PrefixMeta' that automatically adds a prefix "attr_"
# to all user-defined attributes in the class.
# (Hint: modify the 'attrs' dictionary in __new__)
# class PrefixMeta(type):
#     ...

# TODO 3: Create a class 'Simple' that uses PrefixMeta and define a variable 'test = 42'.
# Verify that it is accessible as 'Simple.attr_test'.
# class Simple(metaclass=PrefixMeta):
#     test = 42

# TODO 4: Write a metaclass 'Singleton' that ensures only ONE instance
# of a class can ever be created.
# (Hint: override __call__ in the metaclass to return a stored instance)
# class Singleton(type):
#     ...

# TODO 5: Create a class 'DatabaseConnection' that uses the Singleton metaclass.
# class DatabaseConnection(metaclass=Singleton):
#     ...

# TODO 6: Check if two instances of DatabaseConnection are the same object.

# CHALLENGE: Create a metaclass 'TraceMeta' that wraps all methods of a class
# to print "Entering method: {name}" before execution.

# --- Verification ---
print("--- Results ---")
try:
    # Test your code here
    pass
except NameError as e:
    print(f"Error: {e} — fill in all TODOs")
except Exception as e:
    print(f"Runtime error: {e}")
